cosmos.go

```diff
@@ -266,23 +266,31 @@ func (c *cosmosImpl) ExecuteWithBindings(query string, bindings, rebindings map[
 
 type retryFun func() ([]interfaces.Response, error)
 
-func (c *cosmosImpl) retryLoop(executeRequest retryFun) (responses []interfaces.Response, err error) {
+func retryLoop(executeRequest retryFun, maxRetries int, retryTimeout time.Duration, metrics *Metrics, logger zerolog.Logger) (responses []interfaces.Response, err error) {
+	if metrics == nil {
+		return nil, fmt.Errorf("metrics must not be nil")
+	}
+
 	var tryCount int
-	shouldRetry := c.maxRetries > 0
-	maxTries := c.maxRetries + 1
+	shouldRetry := maxRetries > 0
+	maxTries := maxRetries + 1
 
-	done := make(chan bool)
+	done := make(chan struct{})
 	defer close(done)
 
-	timeoutReachedChan := c.handleTimeout(done)
+	timeoutReachedChan := handleTimeout(done, retryTimeout, logger)
 
 	for tryCount = 0; tryCount < maxTries; tryCount++ {
-
 		responses, err = executeRequest()
+		if err != nil {
+			metrics.requestErrorsTotal.Inc()
+			return nil, errors.Wrap(err, "executing request in retry loop")
+		}
 
-		updateRequestMetrics(responses, c.metrics)
+		isARetry := tryCount > 0
```

Comment:
```suggestion
		isARetry := tryCount > 1
```
TryCount is 0-based in the for-loop